specs/ProgrammaticSaveAs.md

```diff
@@ -0,0 +1,267 @@
+bool ScenarioSaveAs::ProgrammaticSaveAs()
+{
+    if (!m_webView2Staging20)
+        return false;
+    m_webView2Staging20->SaveContentAs(
+        Callback<ICoreWebView2StagingSaveContentAsCompletedHandler>(
+            [this](HRESULT errorCode, COREWEBVIEW2_SAVE_CONTENT_AS_RESULTS result) -> HRESULT
+            {
+                // Show SaveContentAs returned result, optional
+                // Skip message box when the result is COREWEBVIEW2_SAVE_AS_OPEN_SYSTEM_DIALOG (0),
+                // to prevent it deactivates save as modal dialog focus on the window
+                if (result > 0)
+                {
+                    MessageBox(
+                        m_appWindow->GetMainWindow(),
+                        (L"Save As " + saveAsResultString[result]).c_str(), L"Info", MB_OK);
+                }
+                return S_OK;
+            })
+            .Get());
+    return true;
+}
+//! [ProgrammaticSaveAs]
+```
+# API Details
+## Win32 C++
+```c++
+/// Specifies save as type selection options for `ICoreWebView2Staging20`,
+/// used in `SaveAsRequestedEventArgs`
+///
+/// When the source is a html page, allows to select `HTML_ONLY`,
+/// `SINGLE_FILE`, `COMPLETE`; when the source is a non-html,
+/// only allows to select `DEFAULT`; otherwise, will deny the download
+/// and return `COREWEBVIEW2_SAVE_AS_TYPE_NOT_SUPPORTED`.
+///
+/// The content type/format is a MIME type, indicated by the source
+/// server side and identified by the browser. It’s not related to the
+/// file’s type or extension. MIME type of `text/html`,
+/// `application/xhtml+xml` are considered as html page.
+[v1_enum] typedef enum COREWEBVIEW2_SAVE_AS_TYPE {
+  /// Default to save for a non-html content. If it is selected for a html
+  /// page, it’s same as HTML_ONLY option.
+  COREWEBVIEW2_SAVE_AS_TYPE_DEFAULT,
+  /// Save the page as html
+  COREWEBVIEW2_SAVE_AS_TYPE_HTML_ONLY,
+  /// Save the page as mhtml
+  COREWEBVIEW2_SAVE_AS_TYPE_SINGLE_FILE,
+  /// Save the page as html, plus, download the page related source files in
+  /// a folder
+  COREWEBVIEW2_SAVE_AS_TYPE_COMPLETE,
+} COREWEBVIEW2_SAVE_AS_TYPE;
+
+/// Status of a programmatic save as call, indicates the result
+/// for method `SaveContentAs`
+[v1_enum] typedef enum COREWEBVIEW2_SAVE_CONTENT_AS_RESULTS {
+  /// Programmatically open a system default save as dialog
+  COREWEBVIEW2_SAVE_AS_OPEN_SYSTEM_DIALOG,
+  /// Save as downloading not start as given an invalid path
+  COREWEBVIEW2_SAVE_AS_INVALID_PATH,
+  /// Save as downloading not start as given a duplicate filename and
+  /// replace file not allowed
+  COREWEBVIEW2_SAVE_AS_FILE_ALREADY_EXISTS,
+  /// Save as downloading not start as the `SAVE_AS_TYPE` selection not
+  /// supported because of the content MIME type or system limits
+  COREWEBVIEW2_SAVE_AS_TYPE_NOT_SUPPORTED,
+  /// Cancel the save as request
+  COREWEBVIEW2_SAVE_AS_CANCELLED,
+  /// Save as request complete, the downloading started
+  COREWEBVIEW2_SAVE_AS_STARTED
```

Comment:
Once the SaveContentAs method calls back to indicate the asynchronous operation has completed I would expect that it has finished saving the file to the destination file path and I can now open the file in the file path. Is that not the case? Once the SaveContentAs method calls back, the save as is only starting?

I hope we can rename this to the following and only complete the operation after we finish saving the files:
```suggestion
  COREWEBVIEW2_SAVE_AS_COMPLETED
```

Reply:
So how will an end developer know when the Save As actually completes and the files have been written to the disk?

Reply:
This needs to be explained more in the description section in this document and in the reference docs for the SaveContentAs method. Calling the method SaveAs or SaveContentAs makes it sound like the method will save the content. But it doesn't save the content, it starts a download of the main document.

Names that better describe the behavior might be:

* StartDownloadOfMainDocument
* SaveCopyOfDocument

But that doesn't make it clear that this is related to the Save As UI dialog.